toolchain/semantics/semantics_context.h

```diff
@@ -97,6 +97,16 @@ class SemanticsContext {
   // TODO: This should eventually return a type ID.
   auto CanonicalizeType(SemanticsNodeId node_id) -> SemanticsTypeId;
 
+  // Handles canonicalization of struct types. This may create a new struct type
+  // when it has a new structure, or reference an existing struct type when it
+  // duplicates a prior type.
+  //
+  // Individual struct type fields aren't canonicalized because they may have
+  // name conflicts or other diagnostics during creation, which can use the
+  // parse node.
```

Comment:
Rather than avoiding canonicalizing struct type fields, what about refusing to canonicalize the type subtree that contains an invalid node?

(maybe that doesn't actually make sense...)